allennlp/data/tokenizers/pretrained_transformer_tokenizer.py

```diff
@@ -91,42 +92,140 @@ def __init__(
+        self.single_sequence_token_type_id = None
+
+        seen_dummy_a = False
+        for token_id, token_type_id in zip(
+            dummy_output["input_ids"], dummy_output["token_type_ids"]
+        ):
+            if token_id == dummy_a:
+                if seen_dummy_a:
+                    raise ValueError("Cannot auto-determine the number of special tokens added.")
+                seen_dummy_a = True
+                assert (
+                    self.single_sequence_token_type_id is None
+                    or self.single_sequence_token_type_id == token_type_id
+                ), "multiple different token type ids found for the sequence"
+                self.single_sequence_token_type_id = token_type_id
+                continue
+
+            token = Token(
+                tokenizer_with_special_tokens.convert_ids_to_tokens(token_id),
+                text_id=token_id,
+                type_id=token_type_id,
+            )
+            if not seen_dummy_a:
+                self.single_sequence_start_tokens.append(token)
+            else:
+                self.single_sequence_end_tokens.append(token)
+
+        assert (
+            len(self.single_sequence_start_tokens) + len(self.single_sequence_end_tokens)
+        ) == self.tokenizer.num_special_tokens_to_add(pair=False)
+
+    @overrides
+    def tokenize(self, text: str) -> List[Token]:
         """
-        This method works on both sentence and sentence pair.
+        This method only handles a single sentence (or sequence) of text.
         """
-
         encoded_tokens = self.tokenizer.encode_plus(
-            text=sentence_1,
-            text_pair=sentence_2,
-            add_special_tokens=self._add_special_tokens,
+            text=text,
+            add_special_tokens=False,
             max_length=self._max_length,
             stride=self._stride,
             truncation_strategy=self._truncation_strategy,
             return_tensors=None,
+            return_offsets_mapping=self.tokenizer.is_fast,
+            return_attention_mask=False,
             return_token_type_ids=True,
         )
         # token_ids contains a final list with ids for both regular and special tokens
-        token_ids, token_type_ids = encoded_tokens["input_ids"], encoded_tokens["token_type_ids"]
-
-        tokens = []
-        for token_id, token_type_id in zip(token_ids, token_type_ids):
-            token_str = self.tokenizer.convert_ids_to_tokens(token_id, skip_special_tokens=False)
-            tokens.append(Token(text=token_str, text_id=token_id, type_id=token_type_id))
+        token_ids, token_type_ids, token_offsets = (
+            encoded_tokens["input_ids"],
+            encoded_tokens["token_type_ids"],
+            encoded_tokens.get("offset_mapping"),
+        )
 
-        if self._calculate_character_offsets:
+        # If we don't have token offsets, try to calculate them ourselves.
+        if token_offsets is None:
```

Comment:
Maybe make this a separate method?

Reply:
![OdTc3Ce](https://user-images.githubusercontent.com/920638/81749889-565e9880-9461-11ea-9e23-ba8a1cc76d04.png)